Library/Homebrew/cmd/update.rb

```diff
@@ -46,13 +48,53 @@ def update
 
     # automatically tap any migrated formulae's new tap
     report.select_formula(:D).each do |f|
-      next unless (HOMEBREW_CELLAR/f).exist?
+      next unless (dir = HOMEBREW_CELLAR/f).exist?
       migration = TAP_MIGRATIONS[f]
       next unless migration
       tap_user, tap_repo = migration.split "/"
       install_tap tap_user, tap_repo
+      # update tap for each Tab
+      tabs = dir.subdirs.each.map { |d| Tab.for_keg(Keg.new(d)) }
+      tabs.each { |tab| tab.source["tap"] = "#{tap_user}/homebrew-#{tap_repo}" }
+      tabs.each(&:write)
     end if load_tap_migrations
 
+    # Migrate installed renamed formulae from main Homebrew repository.
+    if load_formula_renames
+      report.select_formula(:D).each do |oldname|
+        newname = FORMULA_RENAMES[oldname]
+        next unless newname
+        next unless (dir = HOMEBREW_CELLAR/oldname).directory? && !dir.children.empty?
+        f = Formulary.factory("homebrew/homebrew/#{newname}")
+
+        begin
+          migrator = Migrator.new(f)
+          migrator.migrate
+        rescue Migrator::MigratorDifferentTapsError
+        end
+      end
+    end
+
+    # Migrate installed renamed formulae from taps
+    Tap.each do |tap|
+      unless tap.formula_renames.empty?
+        report.select_formula(:D).each do |oldname|
+          user, repo, oldname = oldname.split("/")
+          next unless user && repo && oldname
+          next unless user == tap.user && repo == tap.repo
+          newname = tap.formula_renames[oldname]
+          next unless newname
+          next unless (dir = HOMEBREW_CELLAR/oldname).directory? && !dir.children.empty?
+
+          begin
+            migrator = Migrator.new(Formulary.factory("#{user}/#{repo}/#{newname}"))
+            migrator.migrate
+          rescue Migrator::MigratorDifferentTapsError
+          end
+        end
+      end
+    end
```

Comment:
The whole block has very poor performance. You should not repeatedly enumerate `report.select_formula`.